### PR TITLE
VSCode: added patchelf after download for NixOS support

### LINKED ISSUE #4542: VSCode extension stopped working on NixOS
**Steps to reproduce:** 
Install any version of extension past 0.2.~143~151, try running any Rust project.

**What I get:**
It doesn't start. When I try to do "Rust Analyzer: Run" (or any other command) it gives me the following message:
image

It worked in Nix up until 0.2.~143~151.

### PR DESCRIPTION
**Pull request by cab404**
This adds Nix support, and fixes #4542

### CODE REVIEW

File: editors/code/src/main.ts [hunk 1]
```diff
@@ -188,6 +189,46 @@ async function bootstrapServer(config: Config, state: PersistentState): Promise<
...
+                exec(`nix-build ${patch_path} --arg src '${orig_file}' -o ${dest}`,
+                (err, stdout, stderr) => {
+                    if (err != null) {
+                        reject(Error(stderr))
+                    } else {
+                        resolve(stdout)
+                    }
+                })
+            })
+            // await fs.unlink(orig_file)
```
Comment (lnicola) [inline editors/code/src/main.ts:227]: Is this needed?

Comment (cab404) [reply editors/code/src/main.ts:227]: yeah, forgot to uncomment

File: editors/code/src/main.ts [hunk 2]
```diff
@@ -188,6 +189,46 @@ async function bootstrapServer(config: Config, state: PersistentState): Promise<
     return path;
 }
 
+async function patchelf(dest: PathLike): Promise<void> {
+    await vscode.window.withProgress(
+        {
+            location: vscode.ProgressLocation.Notification, 
+            title: "Patching rust-analysis for NixOS"
```
Comment (kjeremy) [inline editors/code/src/main.ts:196]:
```suggestion
            title: "Patching rust-analyzer for NixOS"
```

File: editors/code/src/main.ts [hunk 3]
```diff
@@ -237,6 +278,12 @@ async function getServer(config: Config, state: PersistentState): Promise<string
     assert(!!artifact, `Bad release: ${JSON.stringify(release)}`);
 
     await download(artifact.browser_download_url, dest, "Downloading rust-analyzer server", { mode: 0o755 });
+
+    // Patching executable if that's NixOS.
+    if (fs.stat("/etc/nixos").then(_ => true).catch(_ => false)) {
```
Comment (matklad) [inline editors/code/src/main.ts:283]: This misses await?

File: editors/code/src/main.ts [hunk 4]
```diff
@@ -188,6 +189,46 @@ async function bootstrapServer(config: Config, state: PersistentState): Promise<
...
+            await fs.writeFile(patchPath, `
+            {src, pkgs ? import <nixpkgs> {}}:
+                pkgs.stdenv.mkDerivation {
+                    name = "rust-analyzer";
+                    inherit src;
+                    phases = [ "installPhase" "fixupPhase" ];
+                    installPhase = "cp $src $out";
+                    fixupPhase = ''
+                    chmod 755 $out
+                    patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" $out
```
Comment (matklad) [inline editors/code/src/main.ts:210]: Shouldn't this declare a dependency on `patchelf`? Or is it always available?

File: editors/code/src/main.ts [hunk 5]
```diff
@@ -188,6 +189,46 @@ async function bootstrapServer(config: Config, state: PersistentState): Promise<
...
+                    '';
+                }
+            `);
```
Comment (matklad) [inline editors/code/src/main.ts:213]: Is there perhaps some way to specify the scritp on the command line, without writing an inermediate file?

File: editors/code/src/main.ts (hunk 4 above)
Comment (cab404) [reply editors/code/src/main.ts:210]: It is always available in mkDerivation.

File: editors/code/src/main.ts (hunk 5 above)
Comment (cab404) [reply editors/code/src/main.ts:213]: Yes there is, but I don't know how I can find path to it in installed extension :(

File: editors/code/src/main.ts (hunk 3 above)
Comment (cab404) [reply editors/code/src/main.ts:283]: Ooops

File: editors/code/src/main.ts (hunk 5 above)
Comment (cab404) [reply editors/code/src/main.ts:213]: I'll try making it not ugly)

Comment (cab404) [reply editors/code/src/main.ts]: This should be alright

**Comment by cab404** [comment]
bump

**Comment by eternaleye** [comment]
Should this not have read `/etc/os-release`, and branched on the `ID` field being `nixos`? That file was introduced exactly as a stable, consistent means to determine distribution identity, version, etc. without needing O(n) detection heuristics.

**Comment by cab404** [comment]
> Should this not have read `/etc/os-release`, and branched on the `ID` field being `nixos`? That file was introduced exactly as a stable, consistent means to determine distribution identity, version, etc. without needing O(n) detection heuristics.

That was a lot easier (even than writing one line parser for os-release or finding one). I needed to know, whether it is NixOS or not, and that was sufficient :shrug:

**Pull request merged**